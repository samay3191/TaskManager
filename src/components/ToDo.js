import React, { useEffect, useState } from "react";
import { useDispatch, useSelector } from "react-redux";
import { addItem } from "../actions/action";
import Button from "../base/Button";
import TextInput from "../base/TextInput";
import { isEmpty } from "../helper/utils";
import ItemsTable from "./ItemsTable";
import ItemConstants from "../constants/itemConstants";

const ToDo = () => {
  const dispatch = useDispatch();
  const [item, setItem] = useState("");
  const [type, setType] = useState(ItemConstants.ALL);
  const [showList, setShowList] = useState([]);
  const { error, list: toDoList } = useSelector((state) => state.toDo);

  const handleAddItem = () => {
    setItem("");
    dispatch(addItem(item.trim()));
  };

  const updateList = () => {
    switch (type) {
      case ItemConstants.ACTIVE:
        return toDoList.filter((i) => i.selected === false);
      case ItemConstants.SELECTED:
        return toDoList.filter((i) => i.selected === true);
      case ItemConstants.ALL:
      default:
        return toDoList;
    }
  };

  useEffect(() => {
    setShowList(updateList());
  }, [type, toDoList]);

  const Filter = ({ text, filterType }) => (
    <div
      className={`cursor-pointer hover:text-orange-600 m-1 ${
        type === filterType ? "text-blue-700" : ""
      }`}
      onClick={() => setType(filterType)}
    >
      {text}
    </div>
  );

  return (
    <div className="bg-blue-200 rounded-lg m-4 p-2 shadow-sm flex flex-col">
      <div className="flex justify-center flex-wrap my-4">
        <TextInput
          value={item}
          onChange={setItem}
          placeHolder="Inset an Item"
        />
        <Button
          text="Submit"
          onClick={handleAddItem}
          disabled={isEmpty(item)}
        />
      </div>
      <div className="p-1 text-center h-8 text-red-700">
        {!isEmpty(error) && <div className="bg-white rounded-md box-border border border-red-700">{`Error: ${error}`}</div>}
      </div>
      <div className="my-4">
        <ItemsTable data={showList} />
      </div>
      <div className="flex justify-center items-center text-xs font-bold px-1">
        <div>ALL({toDoList.length})</div>
        <div className="mx-auto flex">
          <Filter text="All" filterType={ItemConstants.ALL} />
          <Filter text="Active" filterType={ItemConstants.ACTIVE} />
          <Filter text="Completed" filterType={ItemConstants.SELECTED} />
        </div>
      </div>
    </div>
  );
};

export default ToDo;
