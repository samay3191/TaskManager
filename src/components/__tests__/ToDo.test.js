import React, { Component } from "react";
import { mount, shallow } from "enzyme";
import configureStore from "redux-mock-store";
import ToDo from "../ToDo";
import { Provider } from "react-redux";

const mockStore = configureStore([]);

let component, store;

describe("<ToDo />", () => {
  beforeAll(() => {
    store = mockStore({
      toDo: {
        list: [],
        error: "",
      },
    });
    store.dispatch = jest.fn();
    component = mount(
      <Provider store={store}>
        <ToDo />
      </Provider>
    );
  });

  it("should render ToDo component", () => {
    component.debug();
    expect(component).toMatchSnapshot();
  });
});
