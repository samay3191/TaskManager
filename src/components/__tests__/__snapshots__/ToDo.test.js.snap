// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ToDo /> should render ToDo component 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [MockFunction],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <ToDo>
    <div
      className="bg-blue-200 rounded-lg m-4 p-2 shadow-sm flex flex-col"
    >
      <div
        className="flex justify-center flex-wrap my-4"
      >
        <TextInput
          onChange={[Function]}
          placeHolder="Inset an Item"
          value=""
        >
          <div
            className="m-1 leading-3"
          >
            <input
              className="py-1 border border-gray-400 box-border outline-none rounded-lg px-2"
              onChange={[Function]}
              placeholder="Inset an Item"
              type="text"
              value=""
            />
          </div>
        </TextInput>
        <Button
          disabled={true}
          onClick={[Function]}
          text="Submit"
        >
          <div
            className="m-1 w-12"
          >
            <button
              className="bg-blue-800 h-full text-xs text-white rounded-lg py-2 px-3 leading-3 align-middle opacity-25"
              disabled={true}
              onClick={[Function]}
              type="button"
            >
              Submit
            </button>
          </div>
        </Button>
      </div>
      <div
        className="p-1 text-center h-8 text-red-700"
      />
      <div
        className="my-4"
      >
        <ItemsTable
          data={Array []}
        >
          <div
            className="flex flex-col-reverse"
          />
        </ItemsTable>
      </div>
      <div
        className="flex justify-center items-center text-xs font-bold px-1"
      >
        <div>
          ALL(
          0
          )
        </div>
        <div
          className="mx-auto flex"
        >
          <Filter
            filterType="ALL"
            text="All"
          >
            <div
              className="cursor-pointer hover:text-orange-600 m-1 text-blue-700"
              onClick={[Function]}
            >
              All
            </div>
          </Filter>
          <Filter
            filterType="ACTIVE"
            text="Active"
          >
            <div
              className="cursor-pointer hover:text-orange-600 m-1 "
              onClick={[Function]}
            >
              Active
            </div>
          </Filter>
          <Filter
            filterType="SELECTED"
            text="Completed"
          >
            <div
              className="cursor-pointer hover:text-orange-600 m-1 "
              onClick={[Function]}
            >
              Completed
            </div>
          </Filter>
        </div>
      </div>
    </div>
  </ToDo>
</Provider>
`;
